File: go/ValidBST/ValidBST.go
package main

import (
	"fmt"
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	return checkBST(root, math.MinInt, math.MaxInt)
}

// Time -> 0(N) Space -> 0(logN)
func checkBST(root *TreeNode, min, max int) bool {
	if root == nil {
		return true
	}

	if root.Val <= min || root.Val >= max {
		return false
	}

	return checkBST(root.Left, min, root.Val) && checkBST(root.Right, root.Val, max)
}

func main() {
	left := &TreeNode{Val: 1}
	right := &TreeNode{Val: 3}
	root := &TreeNode{Val: 2, Left: left, Right: right}

	fmt.Println(isValidBST(root))
}

File: python/validBST.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


def isValidBST(root: TreeNode | None) -> bool:
    return checkBST(root, float("-inf"), float("inf"))


# Time -> 0(N) Space -> 0(logN)
def checkBST(root: TreeNode | None, min: float, max: float) -> bool:
    if root is None:
        return True

    if root.val <= min or root.val >= max:
        return False

    return checkBST(root.left, min, root.val) and checkBST(root.right, root.val, max)
